src/gadgets/num.rs

```diff
@@ -425,25 +425,15 @@ impl<E: ScalarEngine> Num<E> {
     }
 
     pub fn scale(mut self, scalar: E::Fr) -> Self {
-        let lc = self
-            .lc
-            .0
-            .iter_mut()
-            .fold(LinearCombination::zero(), |acc, (variable, fr)| {
-                fr.mul_assign(&scalar);
-                acc + (*fr, *variable)
-            });
-
-        let value = match self.value {
-            Some(v) => {
-                let mut tmp = v;
-                tmp.mul_assign(&scalar);
-                Some(tmp)
-            }
-            None => None,
-        };
+        for (_variable, fr) in self.lc.0.iter_mut() {
+            fr.mul_assign(&scalar);
+        }
```

Comment:
This applies to the code being replaced as well, but I am not sure we can safely mutate these `Fr`s here. I think it may be that this is safe in our usage (elsewhere) of `scale`, but not in general. I need to think about this more and also look at how this is used. If it turns out we need to mutate for performance and that it is safe in our usage but not in general, then we should put some safeguards in place (at least documentation) to ensure this is not wrongly used.

Reply:
A few more thoughts, after more time spent looking at this:

- My first comment here was based on the commit introducing the `for` loop. On first reading, I failed to realize that you had modified this code in an earlier commit to this branch. So some of my comments can be attributed to that. In fact, the original implementation of `scale` *didn't* mutably borrow `self`. This explains some of my confusion.
- Regarding 'our usage': `scale` is only used in neptune, and the `Num` on which it called has always been cloned. I think this addresses my concerns above *for that usage*.
- Regarding 'general usage', I think the clone mentioned above is forced by the type system, effectively because `Num` does not implement `Copy`.

My tentative conclusion is that this is all fine. However, this conclusion is a little shaky because it's based on the assumption that some pathological cases may be impossible to construct. But I'm not entirely confident this is true. This may just be me glitching about mutability, though. I think you can probably convince me this is fine. Let's discuss.

Reply:
I am now satisfied this is okay.